File: load-test-framework/src/main/java/com/google/pubsub/clients/flow_control/OutstandingCountFlowController.java
package com.google.pubsub.clients.flow_control;

import com.google.common.cache.Cache;
import com.google.common.cache.CacheBuilder;
import com.google.common.util.concurrent.Monitor;

import java.util.concurrent.Executors;
import java.util.concurrent.ScheduledExecutorService;
import java.util.concurrent.TimeUnit;
import java.util.concurrent.atomic.AtomicInteger;

/**
 * A FlowController that tries to ensure that the outstanding count is roughly equivalent to the
 * completion rate in the next two seconds.
 */
public class OutstandingCountFlowController implements FlowController {
    private static final class None {
        static final None NONE = new None();
        private None() {}
    }

    private int outstanding = 0;
    private double ratePerSecond;
    private final Monitor monitor = new Monitor();
    private final Monitor.Guard requestAllowed = new Monitor.Guard(monitor) {
        // Allow if there are less than 2 seconds of throughput outstanding.
        @Override
        public boolean isSatisfied() {
            return outstanding < (ratePerSecond * 2);
        }
    };
    // ExpiryCache is a cache of counter value to None, used only for the expiration parameter to get the average
    // throughput.
    private final Cache<Integer, None> expiryCache;
    private final AtomicInteger entryIncrementer = new AtomicInteger(0);
    // UpdateExecutor updates the rate at predictable intervals
    private static final int expiryLatencySeconds = 15;
    private static final int rateUpdateDelayMilliseconds = 100;
    private final ScheduledExecutorService updateExecutor = Executors.newSingleThreadScheduledExecutor();

    public OutstandingCountFlowController(double startingPerSecond) {
        this.ratePerSecond = startingPerSecond;
        this.expiryCache = CacheBuilder.newBuilder().expireAfterWrite(expiryLatencySeconds, TimeUnit.SECONDS).build();
        updateExecutor.scheduleAtFixedRate(
                this::updateRate, expiryLatencySeconds * 1000, rateUpdateDelayMilliseconds, TimeUnit.MILLISECONDS);
    }

    private void updateRate() {
        monitor.enter();
        ratePerSecond = expiryCache.size() * 1.0 / expiryLatencySeconds;
        monitor.leave();
    }

    /**
     * Request starting a flow controlled action.
     */
    public void requestStart() {
        monitor.enterWhenUninterruptibly(requestAllowed);
        ++outstanding;
        monitor.leave();
    }

    /**
     * Inform the FlowController that a publish has finished.
     */
    public void informFinished(boolean wasSuccessful) {
        monitor.enter();
        --outstanding;
        monitor.leave();
        if (wasSuccessful) {
            int val = entryIncrementer.incrementAndGet();
            expiryCache.put(val, None.NONE);
        }
    }
}

File: load-test-framework/python_src/clients/flow_control/outstanding_count_flow_controller.py
from clients.flow_control import FlowController
from threading import Condition
from concurrent.futures import ThreadPoolExecutor, Executor
from cachetools import TTLCache
import sys
import time

expiry_latency_seconds = 15
rate_update_delay_seconds = .1


class OutstandingCountFlowController(FlowController):
    """
    A FlowController that tries to ensure that the outstanding count is roughly equivalent to the
    completion rate in the next two seconds.
    """

    def __init__(self, initial_per_second: float):
        self.rate_per_second = initial_per_second
        self.index = 0
        self.outstanding = 0
        self.condition = Condition()
        self.expiry_cache = TTLCache(sys.maxsize, expiry_latency_seconds)
        self.executor: Executor = ThreadPoolExecutor(max_workers=1)
        self.executor.submit(self.reset_rate_in, expiry_latency_seconds)

    def reset_rate_in(self, delay_seconds: float):
        time.sleep(delay_seconds)
        with self.condition:
            self.expiry_cache.expire()
            self.rate_per_second = float(self.expiry_cache.currsize) / expiry_latency_seconds
            self.condition.notify()
        self.reset_rate_in(rate_update_delay_seconds)

    def _num_available(self):
        return int((self.rate_per_second * 2) - self.outstanding)

    def request_start(self):
        with self.condition:
            num_available = self._num_available()
            while num_available < 1:
                self.condition.wait()
                num_available = self._num_available()
            self.outstanding += num_available
            return num_available

    @staticmethod
    def _next_index(index: int):
        return (index + 1) % sys.maxsize

    def inform_finished(self, was_successful: bool):
        with self.condition:
            if was_successful:
                index = self.index
                self.index = self._next_index(index)
                self.expiry_cache[index] = None
            self.outstanding -= 1
            self.condition.notify()
